File: src/providers/CodexDictationClient.cpp
#include "providers/CodexDictationClient.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkRequest>
#include <QTimer>

#include <utility>

#ifdef SPEECHER_WITH_QT_WEBSOCKETS
#include <QWebSocketHandshakeOptions>
#endif

namespace speecher {
namespace {

QString errorMessage(const QJsonObject &event)
{
    return event.value(QStringLiteral("error"))
        .toObject()
        .value(QStringLiteral("message"))
        .toString();
}

bool isAuthenticationError(const QString &message, const QJsonObject &event = {})
{
    const QJsonValue code = event.value(QStringLiteral("error"))
                                .toObject()
                                .value(QStringLiteral("code"));
    const QString codeText = code.toVariant().toString();
    return codeText == QStringLiteral("401") || codeText == QStringLiteral("403")
        || message.contains(QStringLiteral("401"))
        || message.contains(QStringLiteral("403"))
        || message.contains(QStringLiteral("unauthorized"), Qt::CaseInsensitive)
        || message.contains(QStringLiteral("forbidden"), Qt::CaseInsensitive);
}

} // namespace

CodexDictationClient::CodexDictationClient(QObject *parent)
    : QObject(parent)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    connect(&m_socket, &QWebSocket::connected, this, [this] {
        sendSessionStart(m_socket.property("sampleRateHz").toInt());
    });
    connect(&m_socket, &QWebSocket::textMessageReceived,
            this, &CodexDictationClient::handleTextMessage);
    connect(&m_socket, &QWebSocket::disconnected, this, [this] {
        if (!m_cancelled && !m_sessionClosed && !m_failureEmitted) {
            fail(m_sessionStarted
                     ? QStringLiteral("Codex dictation stream closed before the session finished")
                     : QStringLiteral("Codex dictation stream closed before session.start completed"),
                 true,
                 m_finalizing ? QStringLiteral("finalize")
                              : (m_sessionStarted ? QStringLiteral("streaming")
                                                  : QStringLiteral("connect")));
        }
        emit closed();
    });
    connect(&m_socket, &QWebSocket::errorOccurred, this,
            [this](QAbstractSocket::SocketError) {
                if (m_cancelled || m_failureEmitted || m_sessionClosed) {
                    return;
                }
                const QString detail = m_socket.errorString();
                const bool authentication = isAuthenticationError(detail);
                fail(QStringLiteral("Codex dictation connection failed: %1").arg(detail),
                     !authentication,
                     authentication
                         ? QStringLiteral("authentication")
                         : (m_finalizing ? QStringLiteral("finalize")
                                         : (m_sessionStarted ? QStringLiteral("streaming")
                                                             : QStringLiteral("connect"))));
            });
#endif
}

void CodexDictationClient::start(const QUrl &url,
                                 const QString &accessToken,
                                 int sampleRateHz)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    m_pendingAudio.clear();
    m_sessionStarted = false;
    m_finishRequested = false;
    m_finalizing = false;
    m_sessionClosed = false;
    m_cancelled = false;
    m_failureEmitted = false;
    ++m_sessionId;

    m_socket.setProperty("sampleRateHz", sampleRateHz);
    QWebSocketHandshakeOptions options;
    options.setSubprotocols({QStringLiteral("chatgpt-dictation"),
                             QStringLiteral("openai-bearer.%1").arg(accessToken)});
    QNetworkRequest request(url);
    // The production endpoint is guarded as a browser WebSocket. Match the
    // Linux Codex client's Chromium user agent so the upgrade reaches the
    // dictation service instead of Cloudflare's HTML browser challenge.
    request.setHeader(
        QNetworkRequest::UserAgentHeader,
        QStringLiteral("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
                       "(KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36"));
    m_socket.open(request, options);

    const quint64 sessionId = m_sessionId;
    QTimer::singleShot(10000, this, [this, sessionId] {
        if (sessionId == m_sessionId
            && !m_sessionStarted
            && !m_cancelled
            && !m_failureEmitted) {
            fail(QStringLiteral("Codex dictation stream timed out before session.start completed"),
                 true,
                 QStringLiteral("connect"));
            m_socket.abort();
        }
    });
#else
    Q_UNUSED(url)
    Q_UNUSED(accessToken)
    Q_UNUSED(sampleRateHz)
    emit failed(QStringLiteral("Qt WebSockets support was not built; install Qt6 WebSockets development files and rebuild"),
                false,
                QStringLiteral("protocol"));
#endif
}

void CodexDictationClient::sendSessionStart(int sampleRateHz)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    const QJsonObject vad{
        {QStringLiteral("type"), QStringLiteral("server_vad")},
        {QStringLiteral("threshold"), 0.5},
        {QStringLiteral("prefix_padding_ms"), 300},
        {QStringLiteral("silence_duration_ms"), 500},
    };
    const QJsonObject config{
        {QStringLiteral("input_audio_format"), QStringLiteral("pcm16")},
        {QStringLiteral("sample_rate_hz"), sampleRateHz},
        {QStringLiteral("num_channels"), 1},
        {QStringLiteral("max_buffer_size_bytes"), 4 * 1024 * 1024},
        {QStringLiteral("max_utterance_duration_ms"), 30000},
        {QStringLiteral("session_ttl_ms"), 300000},
        {QStringLiteral("provider_mode"), QStringLiteral("streaming_sse")},
        {QStringLiteral("transcript_delivery_mode"), QStringLiteral("final_only")},
        {QStringLiteral("vad"), vad},
    };
    m_socket.sendTextMessage(QString::fromUtf8(QJsonDocument(QJsonObject{
        {QStringLiteral("type"), QStringLiteral("session.start")},
        {QStringLiteral("config"), config},
    }).toJson(QJsonDocument::Compact)));
#else
    Q_UNUSED(sampleRateHz)
#endif
}

void CodexDictationClient::sendAudio(const QByteArray &pcm)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    if (pcm.isEmpty() || m_finishRequested || m_finalizing || m_sessionClosed) {
        return;
    }
    if (!m_sessionStarted) {
        m_pendingAudio.append(pcm);
        return;
    }
    sendAudioMessage(pcm);
#else
    Q_UNUSED(pcm)
#endif
}

void CodexDictationClient::sendAudioMessage(const QByteArray &pcm)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    m_socket.sendTextMessage(QString::fromUtf8(QJsonDocument(QJsonObject{
        {QStringLiteral("type"), QStringLiteral("audio.append")},
        {QStringLiteral("audio"), QString::fromLatin1(pcm.toBase64())},
    }).toJson(QJsonDocument::Compact)));
#else
    Q_UNUSED(pcm)
#endif
}

void CodexDictationClient::flushPendingAudio()
{
    for (const QByteArray &pcm : std::as_const(m_pendingAudio)) {
        sendAudioMessage(pcm);
    }
    m_pendingAudio.clear();
}

void CodexDictationClient::stop()
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    if (m_finishRequested || m_finalizing || m_sessionClosed || m_cancelled) {
        return;
    }
    m_finishRequested = true;
    if (m_sessionStarted) {
        requestFinalization();
    }
#endif
}

void CodexDictationClient::requestFinalization()
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    if (!m_sessionStarted || m_finalizing || m_sessionClosed || m_failureEmitted) {
        return;
    }
    flushPendingAudio();
    m_finalizing = true;
    m_socket.sendTextMessage(QStringLiteral("{\"type\":\"audio.flush\",\"reason\":\"client\"}"));
    m_socket.sendTextMessage(QStringLiteral("{\"type\":\"session.close\"}"));

    const quint64 sessionId = m_sessionId;
    QTimer::singleShot(8000, this, [this, sessionId] {
        if (sessionId == m_sessionId
            && m_finalizing
            && !m_sessionClosed
            && !m_failureEmitted) {
            fail(QStringLiteral("Codex dictation stream timed out while closing the session"),
                 true,
                 QStringLiteral("finalize"));
            m_socket.abort();
        }
    });
#endif
}

void CodexDictationClient::cancel()
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    m_cancelled = true;
    m_pendingAudio.clear();
    m_socket.abort();
#endif
}

bool CodexDictationClient::isConnected() const
{
    return m_sessionStarted && !m_sessionClosed;
}

void CodexDictationClient::handleTextMessage(const QString &message)
{
#ifdef SPEECHER_WITH_QT_WEBSOCKETS
    QJsonParseError parseError;
    const QJsonDocument document = QJsonDocument::fromJson(message.toUtf8(), &parseError);
    if (parseError.error != QJsonParseError::NoError || !document.isObject()) {
        fail(QStringLiteral("Codex dictation stream returned an invalid event payload"),
             false,
             QStringLiteral("protocol"));
        m_socket.close();
        return;
    }

    const QJsonObject event = document.object();
    const QString type = event.value(QStringLiteral("type")).toString();
    if (type == QStringLiteral("session.started")) {
        if (!m_sessionStarted) {
            m_sessionStarted = true;
            flushPendingAudio();
            emit connected();
            if (m_finishRequested) {
                requestFinalization();
            }
        }
        return;
    }
    if (type == QStringLiteral("session.updated")
        && event.value(QStringLiteral("session")).toObject()
               .value(QStringLiteral("status")).toString() == QStringLiteral("closed")) {
        if (!m_sessionClosed) {
            m_sessionClosed = true;
            emit completed();
            m_socket.close();
        }
        return;
    }
    if (type == QStringLiteral("transcript.final")) {
        const QString text = event.value(QStringLiteral("text")).toString().trimmed();
        if (!text.isEmpty()) {
            emit finalTranscript(text);
        }
        return;
    }
    if (type == QStringLiteral("transcript.failed")) {
        const QJsonObject error = event.value(QStringLiteral("error")).toObject();
        const QString message = error.value(QStringLiteral("message")).toString(
            QStringLiteral("Codex dictation failed to transcribe an utterance"));
        const bool authentication = isAuthenticationError(message, event);
        fail(message,
             !authentication && error.value(QStringLiteral("retryable")).toBool(true),
             authentication ? QStringLiteral("authentication")
                            : (m_finalizing ? QStringLiteral("finalize")
                                            : QStringLiteral("streaming")));
        m_socket.close();
        return;
    }
    if (type == QStringLiteral("session.error")
        && event.value(QStringLiteral("fatal")).toBool()) {
        const QString message = errorMessage(event).isEmpty()
            ? QStringLiteral("Codex dictation session failed")
            : errorMessage(event);
        const bool authentication = isAuthenticationError(message, event);
        fail(message,
             !authentication
                 && event.value(QStringLiteral("error")).toObject()
                        .value(QStringLiteral("retryable")).toBool(true),
             authentication ? QStringLiteral("authentication")
                            : (m_finalizing ? QStringLiteral("finalize")
                                            : QStringLiteral("streaming")));
        m_socket.close();
    }
#else
    Q_UNUSED(message)
#endif
}

void CodexDictationClient::fail(const QString &message,
                                bool retryable,
                                const QString &phase)
{
    if (m_cancelled || m_failureEmitted || m_sessionClosed) {
        return;
    }
    m_failureEmitted = true;
    emit failed(message, retryable, phase);
}

} // namespace speecher

File: android/protocol/src/main/kotlin/app/speecher/protocol/CodexDictationClient.kt
package app.speecher.protocol

import java.util.Base64
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive

/** chatgpt.com's speech endpoints expect a browser, as the ChatGPT web app is their client. */
internal const val codexBrowserUserAgent =
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36"

class CodexDictationClient(
    private val transport: WebSocketTransport,
    token: String,
    private val events: (SpeechEvent) -> Unit,
    endpoint: String = "wss://chatgpt.com/backend-api/dictation/stream",
) : WebSocketTransport.Listener, SpeechClient {
    private val lock = Any()
    private val pending = mutableListOf<ByteArray>()
    private val finalIds = mutableSetOf<String>()
    @Volatile private var started = false
    @Volatile private var stopped = false
    @Volatile private var cancelled = false
    @Volatile private var completed = false
    @Volatile private var failed = false
    private val deadline = Executors.newSingleThreadScheduledExecutor { task ->
        Thread(task, "codex-dictation-deadline").apply { isDaemon = true }
    }

    init {
        transport.open(
            endpoint,
            mapOf("User-Agent" to codexBrowserUserAgent),
            "chatgpt-dictation, openai-bearer.$token",
            this,
        )
        deadline.schedule(
            { if (!started) fail(false, "no session.start in 10s") },
            10,
            TimeUnit.SECONDS,
        )
    }

    override fun onOpen() {
        if (cancelled || failed) return
        val vad = buildJsonObject {
            put("type", JsonPrimitive("server_vad"))
            put("threshold", JsonPrimitive(0.5))
            put("prefix_padding_ms", JsonPrimitive(300))
            put("silence_duration_ms", JsonPrimitive(500))
        }
        val config = buildJsonObject {
            put("input_audio_format", JsonPrimitive("pcm16"))
            put("sample_rate_hz", JsonPrimitive(16000))
            put("num_channels", JsonPrimitive(1))
            put("max_buffer_size_bytes", JsonPrimitive(4 * 1024 * 1024))
            put("max_utterance_duration_ms", JsonPrimitive(30000))
            put("session_ttl_ms", JsonPrimitive(300000))
            put("provider_mode", JsonPrimitive("streaming_sse"))
            put("transcript_delivery_mode", JsonPrimitive("segment"))
            put("vad", vad)
        }
        transport.sendText(
            buildJsonObject {
                put("type", JsonPrimitive("session.start"))
                put("config", config)
            }
                .toString()
        )
    }

    override fun onText(text: String) {
        if (cancelled || completed || failed) return
        val event = runCatching {
            Json.parseToJsonElement(text) as JsonObject
        }
            .getOrElse {
                fail(false, "bad event payload")
                return
            }
        when (event.string("type")) {
            "session.started" -> {
                if (started) return
                // Flushed under the lock so audio captured meanwhile queues behind it, not ahead.
                synchronized(lock) {
                    started = true
                    pending.forEach(::sendAudioMessage)
                    pending.clear()
                }
                if (stopped) closeSession() else events(SpeechEvent.Connected)
            }
            "transcript.segment",
            "transcript.final" -> {
                val id = event.string("utterance_id")
                if (id.isNotEmpty() && id in finalIds) return
                val value = event.string("text").trim()
                if (value.isNotEmpty()) {
                    if (event.string("type") == "transcript.final") {
                        if (id.isNotEmpty()) finalIds.add(id)
                        events(SpeechEvent.Final(value))
                    } else events(SpeechEvent.Partial(value))
                }
            }
            "session.updated" ->
                if ((event["session"] as? JsonObject)?.string("status") == "closed") {
                    completed = true
                    deadline.shutdownNow()
                    events(SpeechEvent.Completed)
                    transport.close(1000, null)
                }
            "transcript.failed" -> failWith(event)
            "session.error" -> if (event["fatal"]?.jsonPrimitive?.content == "true") failWith(event)
        }
    }

    override fun onFailure(error: Throwable, statusCode: Int?) {
        val code = statusCode
        val detail =
            listOfNotNull(code?.let { "HTTP $it" }, error.message?.takeIf { it.isNotBlank() })
                .joinToString(": ")
                .ifEmpty { "connect failed" }
        val authentication = code == 401 || code == 403
        fail(authentication, detail, retryable = !authentication)
    }

    override fun onClosed(code: Int, reason: String) {
        fail(false, "closed $code ${reason.take(80)}".trim(), isRetryableClose(code))
    }

    /** A provider error is retryable unless it is about sign-in or says `"retryable": false`. */
    private fun failWith(event: JsonObject) {
        val authentication = event.authenticationError()
        val retryable = (event["error"] as? JsonObject)?.get("retryable")?.jsonPrimitive?.content
        fail(authentication, event.errorDetail(), !authentication && retryable != "false")
    }

    private fun JsonObject.errorDetail(): String {
        val error = this["error"] as? JsonObject ?: return "provider error"
        return "${error.string("code")} ${error.string("message")}".trim().take(120)
    }

    override fun sendAudio(pcm: ByteArray) {
        if (stopped || cancelled || failed || pcm.isEmpty()) return
        synchronized(lock) { if (started) sendAudioMessage(pcm) else pending.add(pcm) }
    }

    private fun sendAudioMessage(pcm: ByteArray) {
        transport.sendText(
            buildJsonObject {
                put("type", JsonPrimitive("audio.append"))
                put("audio", JsonPrimitive(Base64.getEncoder().encodeToString(pcm)))
            }
                .toString()
        )
    }

    override fun stop() {
        if (stopped || cancelled) return
        stopped = true
        if (started) closeSession()
    }

    private fun closeSession() {
        transport.sendText("{\"type\":\"audio.flush\",\"reason\":\"client\"}")
        transport.sendText("{\"type\":\"session.close\"}")
        deadline.schedule({ if (!completed) fail(false, "no close in 8s") }, 8, TimeUnit.SECONDS)
    }

    override fun cancel() {
        cancelled = true
        deadline.shutdownNow()
        synchronized(lock) { pending.clear() }
        transport.cancel()
    }

    private fun fail(authentication: Boolean, detail: String = "", retryable: Boolean = false) {
        val first =
            synchronized(lock) {
                if (cancelled || completed || failed) false
                else {
                    failed = true
                    true
                }
            }
        if (!first) return
        deadline.shutdownNow()
        events(SpeechEvent.Failed(authentication, detail, retryable))
        transport.cancel()
    }
}

private fun JsonObject.string(key: String): String = this[key]?.jsonPrimitive?.content.orEmpty()

private fun JsonObject.authenticationError(): Boolean {
    val error = this["error"] as? JsonObject
    val detail =
        "${error?.string("code").orEmpty()} ${error?.string("message").orEmpty()}".lowercase()
    return isAuthenticationError(detail)
}
